Add Prometheus compatibility survey blog (#4649)
open-telemetry/opentelemetry.io: content/en/blog/2024/prometheus-compatibility-survey/index.md

---
title: Insights from the Prometheus Compatibility Survey
linkTitle: Prometheus Compatibility Survey
date: 2024-06-10

> author: '[David Ashpole](https://github.com/dashpole) (Google)'

Reviewer: ```suggestion
author: '[David Ashpole](https://github.com/dashpole) (Google)'
issue: https://github.com/open-telemetry/sig-end-user/issues/24
sig: End-User SIG
```

Based on our [new blog post guidelines](https://opentelemetry.io/docs/contributing/blog/) we ask blog contributors to provide a "sponsoring SIG" and an issue (normally from within this repo) that approves this PR. Since we discussed that this blog post is going to come beforehand, we can do it like suggested above.

cSpell:ignore: Ashpole
---

> Prometheus and OpenTelemetry are two of the most active and popular projects in

Reviewer: ```suggestion
[Prometheus](https://prometheus.io/) and OpenTelemetry are two of the most active and popular projects in
```

> the CNCF observability landscape. The two communities have been working together

Reviewer: ```suggestion
the [CNCF observability landscape](https://landscape.cncf.io/guide#observability-and-analysis--observability). The two communities have been working together
```

since the early days of OpenTelemetry to improve the compatibility between the

> two projects. The OpenTelemetry Prometheus working group has been leading this

Reviewer: ```suggestion
two projects. The OpenTelemetry Prometheus SIG has been leading this
```

effort, with active participation from maintainers from both OpenTelemetry and
Prometheus.

> At this point, there is a detailed, experimental specification describing how to

Reviewer: ```suggestion
At this point, there is a [detailed, experimental specification](/docs/specs/otel/compatibility/prometheus_and_openmetrics/) describing how to
```

convert between the OpenTelemetry metrics format and the Prometheus metrics
format. It has been used to implement Prometheus (pull) exporters for
OpenTelemetry SDKs, OTLP export from Prometheus libraries, OTLP ingestion for
the Prometheus server, and the OpenTelemetry Collector's Prometheus Receiver and
Prometheus exporters.

As we consider stabilizing the compatibility specification, we decided to run a
survey with the help of the OpenTelemetry End User SIG to collect feedback on
the current state of compatibility between the projects, and try to understand
user sentiment on a few of the remaining open questions. In particular, we
wanted feedback on the translation for metric names, as we want to be sure
before we make any changes.

> The survey received 86 responses, and contained many helpful pieces of feedback.

Reviewer 2: It would be nice to see what the questions were. Could we share them in a separate doc?

Author: The questions can be seen with the results: https://github.com/open-telemetry/sig-end-user/tree/b5cb097ca529cea62809d5078ef8e30a54ad86b9/end-user-surveys/otel-prom-interoperability

Author: I've linked that from the background section

Than you to everyone that participated!

> ## Overall Takeaways

Reviewer 2: ```suggestion
## Overall takeaways
```

We use sentence case for headings. Please edit all headings accordingly — thanks!

- The best predictors of the "units and underscores" group are Prometheus server
  experts and being an SRE.
- The best predictors of the "no units and dots" group are OpenTelemetry library
  experts and being a developer.

## Who took the survey

Survey respondents were mostly from large (>1000 employees) companies (52%) in
the Technology industry (71%). Respondents were more likely to consider
themselves experts with Prometheus-related topics than with
OpenTelemetry-related topics, and were evenly distributed across roles. Nearly
all respondents (>90%) stored metrics in the Prometheus server or another open
source Prometheus backend, and nearly all use PromQL to query their metrics.

## Sentiment on the Current State

Overall, respondents were neutral on the question of whether OpenTelemetry was
easy to use with Prometheus, and considered the current translation between
OpenTelemetry and Prometheus somewhat confusing. This was consistent regardless
of their opinions on units or delimiters.

## Dots and Underscores

OpenTelemetry [specifies](/docs/specs/semconv/general/attribute-naming/) that
conventions should use dots as the namespace delimiter, and underscores as the

> delimiter between "multi-word-dot-delimited components" (e.g.

Reviewer 2: ```suggestion
delimiter between "multi-word-dot-delimited components" (for example,
```

`http.response.status_code`). On the other hand, Prometheus
[uses underscores](https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels)
as its delimiter.

Currently, when exporting in Prometheus format from an OpenTelemetry SDK, all
dots are changed to underscores to comply with the Prometheus requirements. We
wanted to learn whether OpenTelemetry users who used these exporters preferred
to keep the dots in the original metric name, or liked the consistency with
existing Prometheus metrics of translating to underscores.

Of users who indicated they used OpenTelemetry for metrics, and PromQL as their
query language, 54% preferred keeping the original OpenTelemetry metric name
including dots, and 46% want metric names that match Prometheus conventions with
only underscores.

![Dots vs underscores pie chart](dots-vs-underscores.png)

> When we asked about specific example PromQL queries or alerts, the results

Reviewer 2: For the uninitiated, I'd be great to know:

- What are the difficulties that this difference posed so far.
- How is this being tackled, if at all.

Author: added to the background section

and
[requires this unit suffix](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#unit).
Currently, when exporting in Prometheus format from an OpenTelemetry SDK, the
unit is added as a suffix to the metric name.

Of users who indicated they used OpenTelemetry for metrics, and PromQL as their
query language, 37% thought units should be a required suffix for metric names,
and 46% thought units should not be added to metric names. The final 17%
preferred the unit in the metric name, but didn't think it should be required.

![Units in metric name pie chart](units-in-metric-name.png)

When we asked about specific example PromQL queries or alerts, the results were
somewhat more favorable to including the unit in the metric name compared with
the question above. Around 45% of users only selected queries that included the
unit, and around 28% only selected queries that excluded the unit. The final 27%
selected a mix of queries that included or excluded the unit, indicating they
are likely OK with either approach.

## Trends

### Correlation between Unit and Delimiter Preferences

> Not surprisingly, preferences generally split into two groups: Those that prefer

Reviewer 2: Why it's not surprising? Again, some context could help! If this could be the only possible outcome, we might as well get rid of the introductory adverb.

Author: added to the background.

---
title: Insights from the Prometheus Compatibility Survey
linkTitle: Prometheus Compatibility Survey
date: 2024-06-10
author: '[David Ashpole](https://github.com/dashpole) (Google)'
issue: https://github.com/open-telemetry/sig-end-user/issues/24
sig: End-User SIG
cSpell:ignore: Ashpole
---

[Prometheus](https://prometheus.io/) and OpenTelemetry are two of the most
active and popular projects in the
[CNCF observability landscape](https://landscape.cncf.io/guide#observability-and-analysis--observability).
The two communities have been working together since the early days of
OpenTelemetry to improve the compatibility between the two projects. The
OpenTelemetry Prometheus SIG has been leading this effort, with the active
participation of maintainers from both OpenTelemetry and Prometheus.

At this point, there is a
[detailed, experimental specification](/docs/specs/otel/compatibility/prometheus_and_openmetrics/)
describing how to convert between the OpenTelemetry metrics format and the

> Prometheus metrics format. It has been used to implement Prometheus (pull)

Reviewer 3: Add link for `prometheus metrics format`

Author: done

> exporters for OpenTelemetry SDKs, OTLP export from Prometheus libraries, OTLP

Reviewer 3: Add links for `OpenTelemetry SKDs`, `OTLP exporter`, and `Promethueus libraries`

Author: added links

> ingestion for the Prometheus server, and the OpenTelemetry Collector's

Reviewer 3: Add link for `OTLP receiver`, and `OpenTelemetry Collector`

Author: added links

Prometheus Receiver and Prometheus exporters.

One of the most challenging areas to reconcile is that OpenTelemetry metric
names are changed when exporting to Prometheus. Today, the OpenTelemetry
`http.server.request.duration` metric, with unit `s`, is translated to

> `http_server_request_duration_seconds` in Prometheus. Some users are familar

Reviewer 2: ```suggestion
`http_server_request_duration_seconds` in Prometheus. Some users are familiar
```

with the Prometheus naming conventions, and appreciate the consistency this
translation provides with existing metrics in the Prometheus ecosystem. Other
users are confused when querying for the original OpenTelemetry name does not
return any results.

Prometheus is working on support for UTF-8 characters in metric names as part of
its
[2024 roadmap](https://prometheus.io/blog/2024/03/14/commitment-to-opentelemetry/#support-utf-8-metric-and-label-names),
which potentially allows preserving dots in metric names. To better understand

> what users want their Prometheus query experience to look like, we ran a survey

Reviewer 3: ```suggestion
what users want their Prometheus query experience to look like, [we ran a survey](https://github.com/open-telemetry/sig-end-user/tree/main/end-user-surveys/otel-prom-interoperability)
```

> with the help of the OpenTelemetry End User SIG. Deciding on the default

Reviewer 3: ```suggestion
with the help of the [OpenTelemetry End User SIG](/community/end-user/). Deciding on the default
```

translation approach is one of the last remaining blockers for stabilizing the
compatibility specification.

> The survey received 86 responses, and contained many helpful pieces of feedback.

Reviewer 3: ```suggestion
The survey received [91 responses](https://github.com/open-telemetry/sig-end-user/blob/main/end-user-surveys/otel-prom-interoperability/otel-prom-interoperability-survey.csv), and contained many helpful pieces of feedback.
```

See [CSV results](https://github.com/open-telemetry/sig-end-user/blob/main/end-user-surveys/otel-prom-interoperability/otel-prom-interoperability-survey.csv)

Author: 5 of the results were either spam or duplicates.  (e.g. search the results for "visa").  I've omitted them from my analysis below

Reviewer 3: For whomever looks at the raw data or PDF doc, it may be worth mentioning that clarification for the sake of completeness.

Author: added

> Than you to everyone that participated! The questions and raw results can be

Reviewer 3: ```suggestion
Thank you to everyone that participated! The questions and raw results can be
```

found

> [here](https://github.com/open-telemetry/sig-end-user/tree/b5cb097ca529cea62809d5078ef8e30a54ad86b9/end-user-surveys/otel-prom-interoperability).

Reviewer 3: ```suggestion
[here](https://github.com/open-telemetry/sig-end-user/blob/main/end-user-surveys/otel-prom-interoperability/otel-prom-interoperability-survey.csv).
```

names prefer dots in metric names.

### Group Differences

The best predictors of a preference for units required in the name and changing
dots to underscores were having a role of SRE, and being an expert with the
Prometheus server configuration. For example, 88% of SRE respondents preferred
translating dots to underscores.

The best predictors of a preference for preserving the OpenTelemetry name with
dots, and without units were having the role of developer, and being an expert
with OpenTelemetry libraries. For example, 74% of developers preferred not
having units in metric names.

## Other feedback

The most common challenge for all respondents was the instability of
OpenTelemetry instrumentation, and confusion over the conversion logic.
Respondents who preferred OpenTelemetry's conventions listed Prometheus' current
lack of support for OpenTelemetry concepts (resource, scope, delta temporality,
and unit metadata) as their most significant challenge. Respondents who
preferred Prometheus' conventions listed OpenTelemetry's new concepts as
confusing, and were unhappy that OpenTelemetry had deviated from Prometheus'

> existing conventions.

Reviewer 3: ```suggestion
existing conventions.

# Keep in touch

Thanks again to everyone who participated in the survey! We rely on your feedback to help guide the future development of OpenTelemetry and to ensure it continues to meet your evolving needs. We will post upcoming surveys in the following avenues:

 - [#otel-sig-end-user Slack channel](https://cloud-native.slack.com/archives/C01RT3MSWGZ)
   – you can also reach out to us here!
 - [End user resources page](/community/end-user/)
```

Add a wrap-up section

---
title: Insights from the Prometheus Compatibility Survey
linkTitle: Prometheus Compatibility Survey

> date: 2024-06-10

Reviewer: ```suggestion
date: 2024-07-11
```

@dashpole happy to publish this! Let me know what day works for you

Author: Lets do next Monday (7/15), if that works for you.

OpenTelemetry to improve the compatibility between the two projects. The
OpenTelemetry Prometheus SIG has been leading this effort, with the active
participation of maintainers from both OpenTelemetry and Prometheus.

At this point, there is a
[detailed, experimental specification](/docs/specs/otel/compatibility/prometheus_and_openmetrics/)
describing how to convert between the
[OpenTelemetry metrics data model](/docs/specs/otel/metrics/data-model/#opentelemetry-protocol-data-model)
and
[Prometheus metric formats](https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md).
It has been used to implement Prometheus
[(pull) exporters for OpenTelemetry SDKs](https://pkg.go.dev/go.opentelemetry.io/otel/exporters/prometheus),
[OTLP export from Prometheus libraries](https://prometheus.github.io/client_java/otel/otlp/),
[OTLP ingestion for the Prometheus server](https://prometheus.io/docs/prometheus/latest/feature_flags/#otlp-receiver),
and the OpenTelemetry Collector's
[Prometheus Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver)
and

> [Prometheus exporters](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusremotewriteexporter).

Reviewer 4: The link and the description don't seem to match. This links to "Prometheus Remote Write Exporter" instead of "Prometheus exporters", which one do we intend to have here?

Author: By "The OpenTelemetry Collector's ... Prometheus exporters", I was to referencing both the collector's PRW and Prometheus exporters.  I split out PRW and prometheus (pull) exporter to clarify.

> [Prometheus Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver)
> and
> [Prometheus exporters](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusremotewriteexporter).

Author: ```suggestion
[Prometheus Receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver),
[Prometheus Remote Write exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusremotewriteexporter),
and 
[Prometheus (pull) exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusexporter).
```

---
title: Insights from the Prometheus Compatibility Survey
linkTitle: Prometheus Compatibility Survey

> date: 2024-07-11

Reviewer: ```suggestion
date: 2024-07-25
```
